File: day6_part2_2023.rb

def calculate_ways_to_win_long_race(time, record)
  ways_to_win = 0
  (1...time).each do |hold_time|
    travel_time = time - hold_time
    distance = hold_time * travel_time
    ways_to_win += 1 if distance > record
  end
  ways_to_win
end

file = File.open("input.txt")
time, distance = 0, 0

file.each_line do |line|
  next if line.strip.empty?
  parts = line.split(":")
  value = parts[1].delete(" ").to_i
  time == 0 ? time = value : distance = value
end

ways_to_win = calculate_ways_to_win_long_race(time, distance)

puts ways_to_win

File: day6_part2_2023.dart
import 'dart:io';

int calculateWaysToWinLongRace(int time, int record) {
  int waysToWin = 0;
  for (int holdTime = 1; holdTime < time; holdTime++) {
    int travelTime = time - holdTime;
    int distance = holdTime * travelTime;
    if (distance > record) {
      waysToWin++;
    }
  }
  return waysToWin;
}

void main() {
  File file = File('input.txt');
  List<String> lines = file.readAsLinesSync();
  int time = 0;
  int distance = 0;

  for (String line in lines) {
    if (line.isNotEmpty) {
      List<String> parts = line.split(':');
      String value = parts[1].replaceAll(' ', '');
      if (time == 0) {
        time = int.parse(value);
      } else {
        distance = int.parse(value);
      }
    }
  }

  int waysToWin = calculateWaysToWinLongRace(time, distance);

  print(waysToWin);
}